src/JustSaying/Messaging/Channels/IMultiplexer.cs

```diff
@@ -0,0 +1,126 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Channels;
+using System.Threading.Tasks;
+using JustSaying.Messaging.MessageHandling;
+using JustSaying.Messaging.Monitoring;
+using Microsoft.Extensions.Logging;
+
+namespace JustSaying.Messaging.Channels
+{
+    internal interface IMultiplexer
+    {
+        Task Start();
+        void ReadFrom(ChannelReader<IQueueMessageContext> reader);
+        IAsyncEnumerable<IQueueMessageContext> Messages();
+    }
+
+    internal class RoundRobinQueueMultiplexer : IMultiplexer, IDisposable
+    {
+        readonly IList<ChannelReader<IQueueMessageContext>> _readers;
+        Channel<IQueueMessageContext> _targetChannel;
+
+        readonly SemaphoreSlim _syncRoot = new SemaphoreSlim(1, 1);
+        readonly ILogger<RoundRobinQueueMultiplexer> _logger;
+
+        private bool _started = false;
+
+        public RoundRobinQueueMultiplexer(ILoggerFactory loggerFactory)
+        {
+            _readers = new List<ChannelReader<IQueueMessageContext>>();
+            _logger = loggerFactory.CreateLogger<RoundRobinQueueMultiplexer>();
+        }
+
+        public void ReadFrom(ChannelReader<IQueueMessageContext> reader)
+        {
+            if (reader == null) throw new ArgumentNullException(nameof(reader));
+
+            _syncRoot.Wait();
+            _readers.Add(reader);
+            _syncRoot.Release();
+
+            reader.Completion.ContinueWith(c => RemoveReader(reader), TaskScheduler.Default);
+        }
+
+        private void RemoveReader(ChannelReader<IQueueMessageContext> reader)
+        {
+            _logger.LogInformation("Received notification to remove reader from multiplexer inputs");
+
+            _syncRoot.Wait();
+            _readers.Remove(reader);
+            _syncRoot.Release();
+        }
+
+        public async Task Start()
+        {
+            if (_started) return;
+            _started = true;
+
+            var channelCapacity = _readers.Count * 10;
+            _targetChannel = Channel.CreateBounded<IQueueMessageContext>(channelCapacity);
```

Comment:
You can set the `SingleWriter = true` option for this channel

Reply:
and `SingleReader` if I understand right

Reply:
Nice, thanks - imagine this will relieve some locking. There's multiple readers here, though. Consumers connect to this channel via the `IAsyncEnumerable` which can be shared.